src/dataflow/src/render/reduce.rs

```diff
@@ -87,20 +87,39 @@ where
             // Our first step is to extract `(key, vals)` from `input`.
             // We do this carefully, attempting to avoid unneccesary allocations
             // that would result from cloning rows in input arrangements.
-            let group_key_clone = group_key.clone();
-            let aggregates_clone = aggregates.clone();
-
-            // Tracks the required number of columns to extract.
-            let mut columns_needed = 0;
-            for key in group_key.iter() {
-                for column in key.support() {
-                    columns_needed = std::cmp::max(columns_needed, column + 1);
-                }
+
+            let input_arity = input.arity();
+
+            // Form an operator for evaluating key expressions.
+            let mut key_mfp = expr::MapFilterProject::new(input_arity)
```

Comment:
could this all be 1 MFP?

Reply:
Yup (there is a comment up-thread about that: https://github.com/MaterializeInc/materialize/pull/5459#issuecomment-769871341). I can drop a TODO into there so that we don't forget, but I thought that sort of perf-golfing wouldn't help the clarity.